proposals/binary-compat-only.md

```diff
@@ -0,0 +1,123 @@
+> - If the declared accessibility of `M` is `private protected`, let `D` be the intersection of the program text of `P` and the program text of `T` and any type derived from `T`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `protected`, let `D` be the union of the program text of `T`and the program text of any type derived from `T`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `internal`, the accessibility domain of `M` is the intersection of the accessibility domain of `T` with the program text of `P`.
+> - If the declared accessibility of `M` is `private`, the accessibility domain of `M` is the program text of `T`.
+
+The goal of these additions is to make it so that members marked with `BinaryCompatOnlyAttribute` are completely inaccessible to any location, they will
+not participate in member lookup, and cannot affect the rest of the program. Consequentely, this means they cannot implement interface members, they cannot
+call each other, and they cannot be overridden (virtual methods), hidden, or implemented (interface members). Whether this is too strict is the subject of
+several open questions below.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+More complex attribute mechanisms make the language more complex.
+
+## Alternatives
+[alternatives]: #alternatives
+
+No other alternatives have presented themselves
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+### Virtual methods and overriding
+
+What do we do when a virtual method is marked as `BinaryCompatOnly`? Overrides in a derived class may not even be in the current assembly, and it could
+be that the user is looking to introduce a new version of a method that, for example, only differs by return type, something that C# does not normally
+allow overloading on. What happens to any overrides of that previous method on recompile? Are they allowed to override the `BinaryCompatOnly` member if
+they're also marked as `BinaryCompatOnly`?
+
+### Use within the same DLL
+
+This proposal states that `BinaryCompatOnly` members are not visible anywhere, not even in the assembly currently being compiled. Is that too strict, or
+do `BinaryCompatAttribute` members need to possibly chain to one another?
+
+### Implicitly implementing interface members
+
+Should `BinaryCompatOnly` members be able to implement interface members? Or should they be prevented from doing so. This would require that, when a user
+wants to turn an implicit interface implementation into `BinaryCompatOnly`, they would additionally have to provide an explicit interface implementation,
+likely cloning the same body as the `BinaryCompatOnly` member as the explicit interface implementation would not be able to see the original member anymore.
```

Comment:
Perhaps we could have something like `BinaryCompatOnlyImplementation(Type)`, where it specifies an interface which should only be considered to be implemented for binary compatibility purposes. This would allow removing interfaces implementations from usage in source, without breaking their usage in binary.

With this setup, you would disallow explicit interface implementations from being marked `BinaryCompatOnly`, unless their definition is also this as I proposed in my other comment; for a non-`BinaryCompatOnly` definition, if the implicit one is marked `BinaryCompatOnly`, then an explicit one would have to be provided.